src/fparser/two/tests/fortran2003/test_primary_r701.py

```diff
@@ -0,0 +1,196 @@
+# Copyright (c) 2019 Science and Technology Facilities Council
+
+# All rights reserved.
+
+# Modifications made as part of the fparser project are distributed
+# under the following license:
+
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met:
+
+# 1. Redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer.
+
+# 2. Redistributions in binary form must reproduce the above copyright
+# notice, this list of conditions and the following disclaimer in the
+# documentation and/or other materials provided with the distribution.
+
+# 3. Neither the name of the copyright holder nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+'''Test Fortran 2003 rule R701 : primary type.
+
+Sub-rules:
+    C701 (R701) The type-param-name shall be the name of a type parameter.
+    C702 (R701) The designator shall not be a whole assumed-size array
+
+Neither C701 nor C702 can be tested here, as they require context of the
+type defined outside of Primary.
+
+'''
+import sys
+
+import pytest
+
+import fparser.two.Fortran2003 as f2003
+import fparser.two.utils
+
+
+def assert_subclass_parse(source, base_type, actual_type=None,
+                          expected_str=None):
+    '''Assert that the given source matches the given ``base_type``
+    and optionally the specific type that it should produce.
+    '''
+    obj = f2003.Primary(source)
+    # Note: Check that the type exists in the possible types, rather than
+    # checking the type is an instance of one of the possible types (and
+    # all of its ancestor types). This is a much more targetted test as a
+    # result.
+    assert type(obj) in possible_subclasses(base_type)
+
+    if actual_type:
+        assert isinstance(obj, actual_type)
+    else:
+        assert isinstance(obj, base_type)
+
+    if expected_str:
+        assert expected_str == str(obj)
+
+
+def possible_subclasses(node_type, _seen=None):
```

Comment:
Some classes don't just use subclasses to define their subclasses, they also (or alternatively) create them within the match method, so those would be missed in this recursive decent. Is it enough to just check for children, or perhaps can we guarantee that only subclasses are used for matching in all relevant children for this particular test? In any case could you please update the docstring (and function name?) to indicate that this is a test that works in certain circumstances (i.e. is not a general "possible subclasses" function)..